Add CONTRIBUTING Guidelines

UnknownSuperficialNight opened the pull request:
This pull request introduces a **CONTRIBUTING** guidelines document to assist potential contributors in understanding how to effectively participate in **Rodio**. By providing guidelines, the is aim to encourage collaboration and ensure that contributions are of higher quality, while also reducing the effort required to understand the codebase.

Next Steps:
Please review the changes and provide any feedback or suggestions for improvement.

dvdsk commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,110 @@
+
+## Quick Start
+
+1. Install [Rust](https://www.rust-lang.org/tools/install)
```
- Clone the main branch
- Create a new branch for your contribution
- Build rodio using cargo build

UnknownSuperficialNight replied:
Think that should be in the Quick Start?

dvdsk replied:
I saw the later section just now. I would remove Install `Rust` though, I think its fair to assume potential contributors to a rust library have rust installed.

UnknownSuperficialNight replied:
Good point :+1:

dvdsk commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,110 @@
+### Adding a New Audio Source or Effect
+
+1. Create a new file in `src/source/`
+2. Implement the `Source` trait to define how audio samples are generated or modified
```
lets add a hint about wrapping inner sources. Maybe point to amplify as the easiest example.

UnknownSuperficialNight replied:
IDK if u want to do that instead of me, you're probably more qualified for that.

dvdsk replied:
I'll get on it :+1:

dvdsk commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,110 @@
+1. Create a new file in `src/source/`
+2. Implement the `Source` trait to define how audio samples are generated or modified
+3. Consider implementing sources like oscillators, noise generators, or effects like amplification, filtering, or distortion
+4. Optimize for performance, especially for real-time processing in effects
```
lets link to the rust performance book https://nnethercote.github.io/perf-book/introduction.html

dvdsk commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,110 @@
+2. Implement the `Source` trait to define how audio samples are generated or modified
+3. Consider implementing sources like oscillators, noise generators, or effects like amplification, filtering, or distortion
+4. Optimize for performance, especially for real-time processing in effects
+5. Add unit tests/benchmarks where applicable to ensure proper functionality and performance
```
maybe switch the order of 4 and 5. 

also maybe work in there something like this:
We recommend adding a test for your feature before adding it, your PR will need it and it makes development significantly easier. Do not worry about performance or code quality when getting your feature to work. When your feature works, congrats! :tada:. Open a draft PR now as we love helping out during refactoring and optimizing. Then refactor add a benchmark and lets get the implementation fast.

UnknownSuperficialNight replied:
Sure, I paraphrased this and worked it in

should we also mention that some people might not know how to write tests, so maybe I should phrase it as a preferred but not enforced thing, as making them add them could add quite the barrier to entry when they could implement it and then work with us to get the tests in (if applicable)

dvdsk replied:
> Sure, I paraphrased this and worked it in

thanks!

> maybe I should phrase it as a preferred but not enforced thing, as making them add them could add quite the barrier to entry when they could implement it and then work with us to get the tests in (if applicable)

yeah your right there. Ill see if I can work in some links to good resources though

dvdsk replied:
you suggested text is perfect

dvdsk commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,110 @@
+
+## Testing
+
+- Write unit tests for each new function (if applicable)
```
I actually dislike unit tests :sweat_smile:. They make it harder to refactor code since all the unit tests start breaking if you do. Given Rusts robust type system I also do not think they add a lot unlike in langs like python. 

Unit tests can be very valuable during development though! I just remove them after.

UnknownSuperficialNight replied:
Then ill mention that they can add them during development if it helps them, but must be removed before the final merge?

dvdsk replied:
yeah lets do that. "must" sounds so strict though, I want to communicate that they should not spend too much time on the unit tests, they may be very rough since they are temporary.

UnknownSuperficialNight replied:
Check the new changes and let me know if it seems OK

dvdsk commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,110 @@
+## Testing
+
+- Write unit tests for each new function (if applicable)
+- Add integration tests for end-to-end scenarios
```
"If possible add an integration tests. Note: normally these do not need to produce any sound (in other words use the sink/stream). If your adding a source check if the samples have changed, are non zero etc. Unfortunately most things in audio can not easily be checked automatically. For those adding an example that does produce sound and listening to it works.

UnknownSuperficialNight replied:
This what you want me to copy and paste in verbatim or paraphrased?

dvdsk replied:
you could, but do feel free to touch it up.

dvdsk commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,110 @@
+
+1. Fork the repository on GitHub
+2. Clone your fork locally: `git clone https://github.com/YOUR_USERNAME/rodio.git`
+3. Create a feature branch: `git checkout -b feature/your-feature-name`
```
very nice didnt see it yet before I made the first comment

UnknownSuperficialNight replied:
:relieved:

dvdsk commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,110 @@
+  - Reference relevant issue numbers (e.g., "Fixes #123")
+  - Separate logical changes into multiple commits
+  - Avoid commits with unrelated changes
+  Example:
```
love it :+1:

UnknownSuperficialNight replied:
:relieved:

dvdsk reviewed (commented):
havent seen the latest changes yet, getting on that now. Overal it looks very nice!

UnknownSuperficialNight commented:
> havent seen the latest changes yet, getting on that now. Overal it looks very nice!

Thanks :relieved:

dvdsk commented on CONTRIBUTING.md:
```diff
@@ -44,8 +43,12 @@ src/:
-- Write unit tests for each new function (if applicable)
-- Add integration tests for end-to-end scenarios
+- Feel free to write temporary unit tests during development if they help you verify functionality
+- These tests can be rough and don't need to be comprehensive - they're just development aids
```
I really like the wording here, it feels welcoming and inclusive!

dvdsk replied:
we just need some instruction's/help for integration tests, those can be hard to write. This one https://github.com/RustAudio/rodio/blob/6f2518eb1f4ad4e542294f37c074e112f2df1502/tests/seek.rs#L119 took me a ton of time. We should not expect contributors to write *that* involved integration tests but it would be very nice if the could.

The wav test is a good example to point to I think: https://github.com/RustAudio/rodio/blob/master/tests/wav_test.rs

dvdsk replied:
took care of it, please take a look let me know what you think

dvdsk commented:
off topic, I made a draft for a announcement for the next version of rodio in which we ask for help in improving rodio's API. Let me know if you have any feedback on it: https://github.com/RustAudio/rodio/blob/master/outreach/v0.20_announcement_and_call_for_userstories.md (I should have made it a pr, silly me). You can make a new issue, post the feedback here or even make a pr with changes. Whatever is easiest for you.

dvdsk commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,116 @@
+1. Create a new file in `src/source/`
+2. Implement the `Source` trait to define how audio samples are generated or modified
+3. Consider implementing sources like oscillators, noise generators, or effects like amplification, filtering, or distortion
+4. Begin with a test for your new feature. This approach ensures your PR is ready and simplifies development. Don't worry about optimizing initially; focus on functionality.
```
somewhere in between here we should add:

> sources that creates sound get a public (factory) function that constructs the source. This is where you document it. Source that apply effects to other sources get a method with default implementation in the `Source` trait, see:  `src/source/mod.rs`. This method should be the only way to apply the effect and is where it is documented.

This is as short as I could get it

dvdsk commented:
I added some stuff, github actions bot stole the credit stuff. So now I know what the "maintainers can edit this pr" checkbox does, is makes github steal my credit :cry:.

dvdsk commented:
we should edit the outreach/announcement <...>.md  once this is merged. This might pull some potential contributors over the line.

UnknownSuperficialNight commented:
> I added some stuff, github actions bot stole the credit stuff. So now I know what the "maintainers can edit this pr" checkbox does, is makes github steal my credit 😢.
> 
> edit: kinda fixed it.

Oof I already made my own version did not realise your version existed so I incorporated yours into mine let me know what u think

dvdsk commented:
> Oof I already made my own version did not realise your version existed

:laughing: 

>  so I incorporated yours into mine let me know what u think

its better then what I had, only feedback I can think of is changing the current `## Testing` to `### Unit Tests` But its okay so too, you can remove my bit.

And then I think we are done right?

UnknownSuperficialNight commented:
Looks about done, if we can think of anything in the future, we can just make a PR and add it.

dvdsk commented:
:tada:
